modules/admin/admin-ui/src/main/resources/web/admin/apps/content-studio/js/app/wizard/ContentWizardPanel.ts

```diff
@@ -1071,6 +1073,22 @@ export class ContentWizardPanel extends api.app.wizard.WizardPanel<Content> {
         });
     }
 
+    // synch persisted content extra data with data from mixins
+    // when rendering form - we may add extra fields from mixins; as this is intended action from XP, not user - it should be present in persisted content
+    private synchPersistedItemWithMixinData(mixinName: MixinName, mixinData: PropertyTree) {
+        var persistedContent = this.getPersistedItem(),
+            extraData = persistedContent.getExtraData(mixinName);
+        if (!extraData) { // ensure ExtraData object corresponds to each step form
+            extraData = new ExtraData(mixinName, mixinData.copy());
+            persistedContent.getAllExtraData().push(extraData);
+        } else {
```

Comment:
We more or less repeat the code that pushes extra data into content in three different methods now: ``doLayoutPersistedItem``, ``synchPersistedItemWithMixinData`` and ``updateMetadataAndMetadataStepForms``. Let's extract it to a separate method with arguments of ``content``, ``mixinName`` and optionally (if empty - create new instance) ``propertyTree``.